Libraries/Opc.Ua.PubSub/Uadp/UdpClientCreator.cs

```diff
@@ -0,0 +1,296 @@
+        /// Parse the url into an IPaddress and port number
+        /// </summary>
+        /// <param name="url"></param>
+        /// <returns>A new instance of <see cref="IPEndPoint"/> or null if invalid URL.</returns>
+        internal static IPEndPoint GetEndPoint(string url)
+        {
+            Uri connectionUri;
+            if (url != null && Uri.TryCreate(url, UriKind.Absolute, out connectionUri))
+            {
+                if (connectionUri.Scheme != "opc.udp")
+                {
+                    Utils.Trace(Utils.TraceMasks.Error, "Invalid Scheme specified in URL: {0}", url);
+                    return null;
+                }
+                if (connectionUri.Port < 0)
+                {
+                    Utils.Trace(Utils.TraceMasks.Error, "Invalid Port specified in URL: {0}", url);
+                    return null;
+                }
+                string hostName = connectionUri.Host;
+                if (hostName.ToLower() == "localhost")
+                {
+                    hostName = "127.0.0.1";
+                }
+
+                IPAddress ipAddress;
+                if (IPAddress.TryParse(hostName, out ipAddress))
+                {
+                    return new IPEndPoint(ipAddress, connectionUri.Port);
+                }
+                try
+                {
+                    IPHostEntry hostEntry = Dns.GetHostEntry(hostName);
+
+                    //you might get more than one IP for a hostname since 
+                    //DNS supports more than one record
+                    foreach(IPAddress address in hostEntry.AddressList)
+                    {
+                        if (address.AddressFamily == AddressFamily.InterNetwork)
+                        {
+                            return new IPEndPoint(address, connectionUri.Port);
+                        }
+                    }
+                }
+                catch(Exception ex)
+                {
+                    Utils.Trace(ex, "Could not resolve host name: {0}", hostName);
+                }
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Creates and returns a list of <see cref="UdpClient"/> created based on configuration options
+        /// </summary>
+        /// <param name="pubSubContext">Is the method called in a publisher context or a subscriber context</param>
+        /// <param name="networkAddressUrl">The configuration object <see cref="NetworkAddressUrlDataType"/>.</param>
+        /// <param name="configuredEndpoint">The configured <see cref="IPEndPoint"/> that will be used for data exchange.</param>
+        /// <returns></returns>
+        internal static List<UdpClient> GetUdpClients(UsedInContext pubSubContext, NetworkAddressUrlDataType networkAddressUrl, IPEndPoint configuredEndpoint)
+        {
+
+            Utils.Trace(Utils.TraceMasks.Information, "networkAddressUrl.NetworkInterface = {0} \n networkAddressUrl.Url = {1} \n configuredEndpoint = {2}",
```

Comment:
convert \n to better build with stringbuilder and appendline

Reply:
fixed.